stdlib/public/core/DoubleWidth.swift.gyb

```diff
@@ -102,97 +124,26 @@ extension DoubleWidth: Numeric {
 
   @_inlineable // FIXME(sil-serialize-all)
   public init?<T : BinaryInteger>(exactly source: T) {
-    // Can't represent a negative 'source' if Base is unsigned
-    guard source >= 0 || DoubleWidth.isSigned else { return nil }
+    // Can't represent a negative 'source' if Base is unsigned.
+    guard DoubleWidth.isSigned || source >= 0 else { return nil }
     
-    // Is 'source' is entirely representable in Low?
+    // Is 'source' entirely representable in Low?
     if let low = Low(exactly: source.magnitude) {
-      if source < (0 as T) {
-        self.init((~0, ~low + 1))
-      } else {
-        self.init((0, low))
-      }
+      self.init(source < (0 as T) ? (~0, ~low &+ 1) : (0, low))
     } else {
-      // At this point we know source's bitWidth > Base.bitWidth, or else
-      // we would've taken the first branch.
+      // At this point we know source.bitWidth > Base.bitWidth, or else we
+      // would've taken the first branch.
       let lowInT = source & T(~0 as Low)
-      let highInT = source >> High.bitWidth
+      let highInT = source >> Low.bitWidth
       
-      let low = Low(lowInT.magnitude)
+      let low = Low(lowInT)
```

Comment:
Is there a possibility of `lowInT` being negative here?

Reply:
No, because `source.bitWidth > Base.bitWidth`, and `lowInT` is the result of `source` masked by `T(~0 as Low)`, which necessarily clears any sign bit.

Reply:
Makes sense!